### DISCUSSION TITLE
Error with arlstm model

### DISCUSSION THREAD

**Discussion by brett-snider** [neuralhydrology/neuralhydrology#205] [Q&A] [answered]
Hi, 

I am running into an error when I try to create an autoregresive LSTM model. I have copied the example config.yml file but for some reason i get the following typing error (and I can't seem to fix it):

Traceback (most recent call last):
  File "C:\Users\sniderb\AppData\Local\miniconda3\envs\nn_hydro\lib\site-packages\numba\core\dispatcher.py", line 423, in _compile_for_args
    error_rewrite(e, 'typing')
  File "C:\Users\sniderb\AppData\Local\miniconda3\envs\nn_hydro\lib\site-packages\numba\core\dispatcher.py", line 364, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Can't unify return type from the following types: array(float32, 1d, C), array(float64, 1d, C)
Return of: IR name '$12return_value.1', type 'array(float32, 1d, C)', location: 
File "..\..\..\..\AppData\Local\miniconda3\envs\nn_hydro\lib\site-packages\neuralhydrology\utils\samplingutils.py", line 694:
def bernoulli_subseries_sampler(
    <source elided>
    if missing_fraction == 0:
        return data

My config is below:

# --- Experiment configurations --------------------------------------------------------------------
experiment_name: test_run

test_basin_file: summary_vars_added/test_hydat.txt
train_basin_file: summary_vars_added/train_hydat.txt
validation_basin_file: summary_vars_added/test_hydat.txt

seed: 111
device: cpu

# --- Validation configuration ---------------------------------------------------------------------
validate_every: 1
validate_n_random_basins: 4
cache_validation_data: True
metrics:
  - NSE
  - KGE
  - Alpha-NSE
  - Beta-NSE

# --- Model configuration --------------------------------------------------------------------------
model: arlstm
head: regression
output_activation: linear

hidden_size: 16
initial_forget_bias: 3
output_dropout: 0.4

per_basin_test_periods_file: train_hydat_dates.pkl
per_basin_train_periods_file: train_hydat_dates.pkl
per_basin_validation_periods_file: train_hydat_dates.pkl

# --- Training configuration -----------------------------------------------------------------------
optimizer: Adam
loss: MSE
learning_rate:
  0: 1e-3
  10: 5e-4
  20: 1e-4
batch_size: 256
epochs: 1
clip_gradient_norm: 1

predict_last_n: 1
seq_length: 30

num_workers: 0
log_interval: 5
log_tensorboard: True
log_n_figures: 2
save_weights_every: 1
save_validation_results: False

# --- Data configurations --------------------------------------------------------------------------
dataset: generic
data_dir: /ungauged_flow_prediction/HYSET/
forcings: # will be filled by conftest
dynamic_inputs:
  - total_precipitation
  - 2m_tasmax
  - 2m_tasmin
  - snow_water_equivalent
target_variables:
  - discharge_mm
static_attributes:
- Centroid_Lat_deg_N
- Centroid_Lon_deg_E
- Drainage_Area_km2
- Elevation_m
- Slope_deg
- Gravelius
- Perimeter
- Aspect_deg
- Land_Use_Forest_frac
- Land_Use_Grass_frac
- Land_Use_Wetland_frac
- Land_Use_Water_frac
- Land_Use_Urban_frac
- Land_Use_Shrubs_frac
- Land_Use_Crops_frac
- Permeability_logk_m2
- Porosity_frac
- Hydrometric_station_latitude
- Hydrometric_station_longitude
- avg_precip
- avg_swe
- high_precip_freq
- low_precip_freq
- high_precip_duration
- low_precip_duration
lagged_features:
 discharge_mm:
   - 1
autoregressive_inputs:
  - discharge_mm_shift1
random_holdout_from_dynamic_features:
  discharge_mm_shift1:
    missing_fraction: 0
    mean_missing_length: 10

**Comment by gauchm** [comment, marked as answer]
I'm guessing, but can you try adding a `dtype=np.float32` to the call of `np.full` [here](https://github.com/neuralhydrology/neuralhydrology/blob/master/neuralhydrology/utils/samplingutils.py#L717)? By default, np.full will return float64, so maybe numba doesn't like that if `data` is float32. 

If this doesn't work, the simplest workaround to unblock you would be to remove the `@njit` annotation in the definition of [bernoulli_subseries_sampler](https://github.com/neuralhydrology/neuralhydrology/blob/master/neuralhydrology/utils/samplingutils.py#L655). But that might make everything a lot slower, so it's not a great solution.
